kong/db/dao/plugins/external.lua

```diff
@@ -0,0 +1,642 @@
+
+        ngx_timer_at(0, function()
+          local co = coroutine.running()
+          save_for_later[co] = saved
+
+          handle_event(server_rpc, self.name, conf, phase)
+
+          save_for_later[co] = nil
+        end)
+      end
+
+    else
+      plugin[phase] = function(self, conf)
+        handle_event(server_rpc, self.name, conf, phase)
+      end
+    end
+  end
+
+  return plugin
+end
+
+
+--[[
+
+Plugin info requests
+
+Disclaimer:  The best way to do it is to have "ListPlugins()" and "GetInfo(plugin)"
+RPC methods; but Kong would like to have all the plugin schemas at initialization time,
+before full cosocket is available.  At one time, we used blocking I/O to do RPC at
+non-yielding phases, but was considered dangerous.  The alternative is to use
+`io.popen(cmd)` to ask fot that info.
+
+
+In the external plugins configuration, the `.info_cmd` field contains a string
+to be executed as a command line.  The output should be a JSON string that decodes
+as an array of objects, each defining the name, priority, version and schema of one
+plugin.
+
+    [{
+      "name": ... ,
+      "priority": ... ,
+      "version": ... ,
+      "schema": ... ,
+      "phases": [ phase_names ... ],
+    },
+    {
+      ...
+    },
+    ...
+    ]
+
+This array should describe all plugins currently available through this server,
+no matter if actually enabled in Kong's configuration or not.
+
+--]]
+
+
+local function register_plugin_info(server_def, plugin_info)
+  if _plugin_infos[plugin_info.Name] then
+    kong.log.err(string.format("Duplicate plugin name [%s] by %s and %s",
+      plugin_info.Name, _plugin_infos[plugin_info.Name].server_def.name, server_def.name))
+    return
+  end
+
+  _plugin_infos[plugin_info.Name] = {
+    server_def = server_def,
+    --rpc = server_def.rpc,
+    name = plugin_info.Name,
+    PRIORITY = plugin_info.Priority,
+    VERSION = plugin_info.Version,
+    schema = plugin_info.Schema,
+    phases = plugin_info.Phases,
+  }
+end
+
+local function ask_info(server_def)
+  if not server_def.query_command then
+    kong.log.info(string.format("No info query for %s", server_def.name))
+    return
+  end
+
+  local fd, err = io.popen(server_def.query_command)
+  if not fd then
+    local msg = string.format("loading plugins info from [%s]:\n", server_def.name)
+    kong.log.err(msg, err)
+    return
+  end
+
+  local infos_dump = fd:read("*a")
+  fd:close()
+  local infos = cjson_decode(infos_dump)
+  if type(infos) ~= "table" then
+    error(string.format("Not a plugin info table: \n%s\n%s",
+        server_def.query_command, infos_dump))
+    return
+  end
+
+  for _, plugin_info in ipairs(infos) do
+    register_plugin_info(server_def, plugin_info)
+  end
+end
+
+local function load_all_infos()
+  if not _plugin_infos then
+    _plugin_infos = {}
+
+    for _, server_def in ipairs(get_server_defs()) do
+      ask_info(server_def)
+    end
+  end
+
+  return _plugin_infos
+end
+
+
+local loaded_plugins = {}
```

Comment:
Why do we need a separate `loaded_plugins` variable? It seems it could be merged with `_plugin_infos` (and perhaps read the phases as part of it)?

Reply:
`_plugin_infos` contains the info of all available plugins, as answered by the pluginservers.  `loaded_plugins` contains all the plugins loaded, in runnable state (with phase closures).  might be merged into one, but i found easier to reason about with the split between "cached external data" and "active object".

i'm also semi-foggily thinking about eventually moving to a more dynamic lifecycle, where plugins could be added/removed at runtime.  these will have to be kept in sync, which could be tricky; but at the same time it might be important to differentiate between the new info from the external pluginserver and the still-running plugin.   hard to say if merging would simplify or complicate.